cmd/registry/cmd/upload/bulk/protos.go

```diff
@@ -289,10 +305,117 @@ func (task *uploadProtoTask) fileName() string {
 
 func (task *uploadProtoTask) zipContents() ([]byte, error) {
 	prefix := task.directory + "/"
-	contents, err := core.ZipArchiveOfPath(task.path, prefix, true)
+
+	// Get the proto files in the main directory.
+	protos, err := localProtos(task.path, prefix)
+	if err != nil {
+		return nil, err
+	}
+
+	// Compile the listed protos to get their dependencies.
+	count := len(protos)
+	for count > 0 {
+		protos, err = referencedProtos(protos, task.directory)
+		if err != nil {
+			return nil, err
+		}
+		if len(protos) == count {
+			break
+		}
+		count = len(protos)
+	}
+
+	// Get the metadata files in the main directory.
+	metadata, err := localMetadata(task.path, prefix)
+	if err != nil {
+		return nil, err
+	}
+
+	allfiles := append(metadata, protos...)
+
+	// Zip the listed files.
+	contents, err := core.ZipArchiveOfFiles(allfiles, prefix, true)
 	if err != nil {
 		return nil, err
 	}
 
 	return contents.Bytes(), nil
 }
+
+// Collect the names of all metadata files in a source directory, stripping the prefix.
+func localMetadata(source, prefix string) ([]string, error) {
+	filenames := make([]string, 0)
+	err := filepath.WalkDir(source, func(p string, entry fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		} else if entry.IsDir() {
+			return nil // Do nothing for the directory, but still walk its contents.
+		} else if strings.HasSuffix(p, ".json") || strings.HasSuffix(p, ".yaml") {
+			filenames = append(filenames, strings.TrimPrefix(p, prefix))
+		}
+		return nil
+	})
+	return filenames, err
+}
+
+// Collect the names of all proto files in a source directory, stripping the prefix.
+func localProtos(source, prefix string) ([]string, error) {
+	filenames := make([]string, 0)
+	err := filepath.WalkDir(source, func(p string, entry fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		} else if entry.IsDir() {
+			return nil // Do nothing for the directory, but still walk its contents.
+		} else if strings.HasSuffix(p, ".proto") {
+			filenames = append(filenames, strings.TrimPrefix(p, prefix))
+		}
+		return nil
+	})
+	return filenames, err
+}
+
+// Get all the protos that are referenced in the compilation of a list of protos.
+func referencedProtos(protos []string, root string) ([]string, error) {
+	tempDir, err := os.MkdirTemp("", "proto-import-")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(tempDir)
+	args := []string{"-o", tempDir + "/proto.pb"}
+	args = append(args, protos...)
+	cmd := exec.Command("protoc", args...)
+	cmd.Dir = root
+	if err := cmd.Run(); err != nil {
+		return nil, fmt.Errorf("failed to compile protos with protoc: %s", err)
+	}
+	filenames, err := protosFromFileDescriptorSet(tempDir + "/proto.pb")
```

Comment:
This can be just `return protosFromFileDescriptorSet(tempDir + "/proto.pb")`.

Reply:
Fixed, thanks!